### ISSUE TITLE
Export documents 

### ISSUE THREAD

**Issue by tendant** [meilisearch/MeiliSearch#550]
There is api to retrieve all documents using documents query with "offset" and "limit". 

However depending on how order is implemented in MeiliSearch. This api might not be able to reliably provide all documents for some given criteria. 

Is there any way to export all documents in one batch?

**Comment by Kerollmops** [comment]
We have [a route `/:index/documents`](https://docs.meilisearch.com/references/documents.html#get-documents) that will returns all documents by batches of a maximum of 1000. But the engine doesn't have currently any other reliable way to "export" the documents.

Does this route fits your needs?

**Comment by tendant** [comment]
I am aware of the /:index/documents api. It is not a reliable way of exporting documents, if there is any new document added during exporting. 

With exporting feature, MeiliSearch can be used as a database in many cases. 

Thanks to MeiliSearch's fast search and easy to use api, it make sense to use it in place of other NoSQL databases.

**Comment by Kerollmops** [comment]
Hey @tendant,

We thought about this feature a little bit and found out that this is a potential performance and memory consumption problem for the engine **and** for the client. This is a highly related to the data format we use, JSON.

If we provide an export route the engine must use a read transaction and create a potentialy enormous array containing all the documents, this big array can make the engine goes out of memory and crash the program.

Another solution would be to create this JSON array on the fly, pushing the first `[` then sending every document by streaming the lines and printing the last `]`, this bring another issue but this time on the client side, an array is not a streamable data structure and therefore the client must collect it in memory before being able to use it, the client can trigger an out of memory.

The last solution would be to use another data format, JSON Lines for example, this way we would be able to keep the memory low while streaming the documents to the client and client can also stream it anywhere it wants without having to way for the entire dataset. The problem is that we use a different format and it is a little bit strange for the users of the API.

Ultimately all these solutions have the same problem, keeping a read transaction for too much time [can trigger significant disk usage growth](http://www.lmdb.tech/doc/group__mdb.html#ga5d51d6130325f7353db0955dbedbc378) if run in parallel with write transactions. But that can be a warning in the documentation.

Note that I would **love** to be able to use MeiliSearch without any additional database.

**Comment by tendant** [comment]
Thanks for the detailed explanation. That makes total sense to me. 

The intend to export documents is mainly to use it for backup or migration. It is not necessary to have it as runtime service. Even just a offline tool or script will do the job. In such case, it will have no impact on the service itself. This is very important for a production deployment.

**Comment by qdequele** [comment]
Closed in favor of #561
